x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/Greatest.java

```diff
@@ -0,0 +1,156 @@
+    private DataType dataType;
+
+    public Greatest(Source source, List<Expression> expressions) {
+        super(source, expressions);
+    }
+
+    @Override
+    public DataType dataType() {
+        if (dataType == null) {
+            resolveType();
+        }
+        return dataType;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        for (int position = 0; position < children().size(); position++) {
+            if (dataType == null || dataType == NULL) {
+                dataType = children().get(position).dataType();
+                continue;
+            }
+            TypeResolution resolution = TypeResolutions.isType(
+                children().get(position),
+                t -> t == dataType,
+                sourceText(),
+                TypeResolutions.ParamOrdinal.fromIndex(position),
+                dataType.typeName()
+            );
+            if (resolution.unresolved()) {
+                return resolution;
+            }
+        }
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    @Override
+    public ScriptTemplate asScript() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new Greatest(source(), newChildren);
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, Greatest::new, children());
+    }
+
+    @Override
+    public boolean foldable() {
+        return Expressions.foldable(children());
+    }
+
+    @Override
+    public Object fold() {
+        return Mappable.super.fold();
+    }
+
+    @Override
+    public Supplier<EvalOperator.ExpressionEvaluator> toEvaluator(
+        Function<Expression, Supplier<EvalOperator.ExpressionEvaluator>> toEvaluator
+    ) {
+        List<Supplier<EvalOperator.ExpressionEvaluator>> evaluatorSuppliers = children().stream().map(toEvaluator).toList();
+        Supplier<EvalOperator.ExpressionEvaluator[]> supplier = () -> evaluatorSuppliers.stream()
+            .map(Supplier::get)
+            .toArray(EvalOperator.ExpressionEvaluator[]::new);
+        if (dataType() == DataTypes.INTEGER) {
+            return () -> new GreatestIntEvaluator(supplier.get());
+        }
+        if (dataType() == DataTypes.LONG) {
+            return () -> new GreatestLongEvaluator(supplier.get());
+        }
+        if (dataType() == DataTypes.DOUBLE) {
+            return () -> new GreatestDoubleEvaluator(supplier.get());
+        }
+        if (DataTypes.isString(dataType())) {
+            return () -> new GreatestBytesRefEvaluator(supplier.get());
+        }
+        throw new UnsupportedOperationException();
+    }
+
+    @Evaluator(extraName = "Int")
+    static int process(int[] values) {
+        int max = values[0];
```

Comment:
nit: can't this be out of bounds? Not sure the case is handled when the number of varargs is zero.

Reply:
The function requires at least one argument. Or, at least, it should. Let me make sure there's a test for that case.

Reply:
Looks like I've made this error a few times!